### PR TITLE
[RTL][KMTEST] RtlPcToFileHeader doesn't handle user mode images in kernel mode

### PR DESCRIPTION
**Pull request by tkreuzer**
Fixes wrong behavior of RtlPcToFileHeader. KM test included.

### CODE REVIEW

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 1]
```diff
@@ -130,7 +130,69 @@ TestStackWalk1(VOID)
...
+
+DECLSPEC_NOINLINE
+static
+VOID
+TestRtlPcToFileHeader(VOID)
+{
+    PVOID ImageBase, Result;
+    PTEB Teb;
+    PPEB Peb;
+    __debugbreak();
```
Comment (learn-more) [inline modules/rostests/kmtests/rtl/RtlStack.c:151]: You probably forgot to remove this?

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 2]
```diff
@@ -130,7 +130,69 @@ TestStackWalk1(VOID)
...
+        ok_eq_pointer(Result, ImageBase);
+#ifdef NTOS_MODE_USER
+        ok_eq_pointer(ImageBase, Peb->ImageBaseAddress);
+#else
+        ok_eq_pointer(ImageBase, NULL);
+#endif
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        ok(FALSE, "Gor exception!");
```
Comment (HBelusca) [inline modules/rostests/kmtests/rtl/RtlStack.c:186]:
```suggestion
        ok(FALSE, "Got exception!");
```

File: modules/rostests/kmtests/rtl/RtlStack.c (hunk 1 above)
Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c:151]: yup

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 3]
```diff
@@ -130,7 +130,69 @@ TestStackWalk1(VOID)
...
+    _SEH2_TRY
+    {
+        Peb = Teb->ProcessEnvironmentBlock;
+        ok(Peb != NULL, "Peb is NULL!");
+        if (Peb == NULL)
+        {
+            return;
+        }
+
+        Result = RtlPcToFileHeader((PUCHAR)Peb->ImageBaseAddress + 0x1000, &ImageBase);
```
Comment (HBelusca) [inline modules/rostests/kmtests/rtl/RtlStack.c:176]: Why 0x1000? Does it show the expected behaviour only when looking at an address at least one page higher than the original base address?

Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c:176]: It's just a random value

Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c:178]: Added a comment.

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 4]
```diff
@@ -130,7 +130,71 @@ TestStackWalk1(VOID)
...
+#ifdef NTOS_MODE_USER
+    Teb = NtCurrentTeb();
+#else
+    Teb = KeGetCurrentThread()->Teb;
+#endif
+    ok(Teb != NULL, "Teb is NULL!");
+    if (Teb == NULL)
+    {
+        return;
+    }
```
Comment (SergeGautherie) [inline modules/rostests/kmtests/rtl/RtlStack.c:166]: Use `if (skip(...))`, both occurrences?

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 5]
```diff
@@ -130,7 +130,71 @@ TestStackWalk1(VOID)
...
+    _SEH2_TRY
+    {
+        Peb = Teb->ProcessEnvironmentBlock;
+        ok(Peb != NULL, "Peb is NULL!");
+        if (Peb == NULL)
+        {
+            return;
+        }
+
+        /* Test an address somewhere within the main image of the current process */
```
Comment (ThFabba) [inline modules/rostests/kmtests/rtl/RtlStack.c:177]: Is there any point in testing how the UM function behaves with a kernel-mode address?

**Review by zefklop — APPROVED** [review]

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 6]
```diff
@@ -130,7 +130,71 @@ TestStackWalk1(VOID)
...
+    Result = RtlPcToFileHeader(&TestRtlPcToFileHeader, &ImageBase);
+    ok_eq_pointer(Result, ImageBase);
+    ok_eq_pointer(ImageBase, &__ImageBase);
+
+#ifdef NTOS_MODE_USER
+    Teb = NtCurrentTeb();
+#else
+    Teb = KeGetCurrentThread()->Teb;
+#endif
+    ok(Teb != NULL, "Teb is NULL!");
```
Comment (ThFabba) [inline modules/rostests/kmtests/rtl/RtlStack.c:162]: Clearly your tests are passing... all our `ok()` calls are missing the `\n` ;)

Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c]: Fixed

Comment (SergeGautherie) [reply modules/rostests/kmtests/rtl/RtlStack.c]: > Fixed

Missed 1 out of 3.

File: ntoskrnl/rtl/libsupp.c [hunk 7]
```diff
@@ -37,6 +37,10 @@ RtlPcToFileHeader(
 {
     PLDR_DATA_TABLE_ENTRY LdrEntry;
     BOOLEAN InSystem;
+    KIRQL OldIrql;
+
+    /* Acquire the loaded module spinlock */
+    OldIrql = KeAcquireSpinLockRaiseToSynch(&PsLoadedModuleSpinLock);
```
Comment (zefklop) [inline ntoskrnl/rtl/libsupp.c:43]: Why raise to synch ?

Comment (tkreuzer) [reply ntoskrnl/rtl/libsupp.c:43]: I guess it's not needed, I simply copied it from MiProcessLoaderEntry. I'll change to dispatch.

File: modules/rostests/kmtests/rtl/RtlStack.c (hunk 5 above)
Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c:177]: I don't think so, but feel free to add it.

File: modules/rostests/kmtests/rtl/RtlStack.c [hunk 8]
```diff
@@ -130,7 +130,71 @@ TestStackWalk1(VOID)
...
+}
+#endif // _M_AMD64
+
 START_TEST(RtlStack)
 {
     TestStackWalk1();
+#ifdef _M_AMD64
+    TestRtlPcToFileHeader();
+#endif // _M_AMD64
 }
```
Comment (binarymaster) [inline modules/rostests/kmtests/rtl/RtlStack.c:200]: Would be nice to have kmtests changes in a separate commit. Or reflect the change in existing commit message, like... 
```
[NTOS:RTL][KMTESTS:RTL] RtlPcToFileHeader doesn't handle user mode images in kernel mode
```

File: modules/rostests/kmtests/rtl/RtlStack.c (hunk 6 above)
Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c]: Fixed

File: modules/rostests/kmtests/rtl/RtlStack.c (hunk 8 above)
Comment (tkreuzer) [reply modules/rostests/kmtests/rtl/RtlStack.c:200]: Updated the commit message

**Pull request merged**